File: examples/FSharpProd/HttpTests/AdvancedHttpWithConfig.fs
module FSharpProd.HttpTests.AdvancedHttpWithConfig
open FSharp.Control.Tasks.NonAffine
open Newtonsoft.Json
open NBomber
open NBomber.Contracts
open NBomber.FSharp
open NBomber.Plugins.Http.FSharp
open NBomber.Plugins.Network.Ping

open FSharpProd.HttpTests.AdvancedHttpTest

// in this example we use:
// - NBomber.Http (https://nbomber.com/docs/plugins-http)

[<CLIMutable>]
type User = {
    Id: int
    Name: string
}

let run () =

    let userFeed = FeedData.fromJson<User> "./HttpTests/Configs/user-feed.json"
                   |> Feed.createRandom "userFeed"

    let getUser = Step.create("get_user", feed = userFeed, execute = fun context ->

        let userId = context.FeedItem
        let url = "https://jsonplaceholder.typicode.com/users?id=" + userId.Id.ToString()

        Http.createRequest "GET" url
        |> Http.withCheck(fun response -> task {
            let! json = response.Content.ReadAsStringAsync()

            // parse JSON
            let users = json
                        |> JsonConvert.DeserializeObject<UserResponse[]>
                        |> ValueOption.ofObj

            match users with
            | ValueSome usr when usr.Length = 1 ->
                return Response.ok(usr.[0]) // we pass user object response to the next step

            | _ -> return Response.fail("not found user: " + userId.Id.ToString())
        })
        |> Http.send context
    )

    // this 'getPosts' will be executed only if 'getUser' finished OK.
    let getPosts = Step.create("get_posts", execute = fun context ->

        let user = context.GetPreviousStepResponse<UserResponse>()
        let url = "https://jsonplaceholder.typicode.com/posts?userId=" + user.Id

        Http.createRequest "GET" url
        |> Http.withCheck(fun response -> task {
            let! json = response.Content.ReadAsStringAsync()

            // parse JSON
            let posts = json
                        |> JsonConvert.DeserializeObject<PostResponse[]>
                        |> ValueOption.ofObj

            match posts with
            | ValueSome ps when ps.Length > 0 ->
                return Response.ok()

            | _ -> return Response.fail()
        })
        |> Http.send context
    )

    Scenario.create "rest_api" [getUser; getPosts]
    |> NBomberRunner.registerScenario
    |> NBomberRunner.withWorkerPlugins [new PingPlugin()]
    |> NBomberRunner.loadConfig "./HttpTests/Configs/config.json"
    |> NBomberRunner.loadInfraConfig "./HttpTests/Configs/infra-config.json"
    |> NBomberRunner.run
    |> ignore

File: examples/CSharpProd/HttpTests/AdvancedHttpWithConfig.cs
using System.Linq;
using NBomber;
using NBomber.Contracts;
using NBomber.CSharp;
using NBomber.Plugins.Http.CSharp;
using NBomber.Plugins.Network.Ping;
using Newtonsoft.Json;

namespace CSharpProd.HttpTests
{
    public class UserId
    {
        public string Id { get; set; }
    }

    public class AdvancedHttpWithConfig
    {
        public static void Run()
        {
            var userFeed = Feed.CreateRandom(
                "userFeed",
                FeedData.FromJson<UserId>("./HttpTests/Configs/user-feed.json")
            );
            var httpFactory = HttpClientFactory.Create();

            var getUser = Step.Create("get_user", httpFactory, userFeed, async context =>
            {
                var userId = context.FeedItem;
                var url = $"https://jsonplaceholder.typicode.com/users?id={userId}";

                var request =  Http.CreateRequest("GET", url)
                    .WithCheck(async response =>
                    {
                        var json = await response.Content.ReadAsStringAsync();

                        // parse JSON
                        var users = JsonConvert.DeserializeObject<UserResponse[]>(json);

                        return users?.Length == 1
                            ? Response.Ok(users.First()) // we pass user object response to the next step
                            : Response.Fail("not found user");
                    });
                var response = await Http.Send(request, context);
                return response;
            });

            // this 'getPosts' will be executed only if 'getUser' finished OK.
            var getPosts = Step.Create("get_posts", httpFactory, async context =>
            {
                var user = context.GetPreviousStepResponse<UserResponse>();
                var url = $"https://jsonplaceholder.typicode.com/posts?userId={user.Id}";

                var request =  Http.CreateRequest("GET", url)
                    .WithCheck(async response =>
                    {
                        var json = await response.Content.ReadAsStringAsync();

                        // parse JSON
                        var posts = JsonConvert.DeserializeObject<PostResponse[]>(json);

                        return posts?.Length > 0
                            ? Response.Ok()
                            : Response.Fail($"not found posts for user: {user.Id}");
                    });
                var response = await Http.Send(request, context);
                return response;
            });

            var scenario = ScenarioBuilder
                .CreateScenario("rest_api", getUser, getPosts);

            NBomberRunner
                .RegisterScenarios(scenario)
                .WithWorkerPlugins(new PingPlugin())
                .LoadConfig("./HttpTests/Configs/config.json")
                .LoadInfraConfig("./HttpTests/Configs/infra-config.json")
                .Run();
        }
    }
}
